### ISSUE TITLE
Order agnostic min/max deterministic computation 

### ISSUE THREAD

**Issue by laithsakka** [facebookincubator/velox#7633]
### Description

One problem with Min/Max computation is that Presto will stop at the first seen null, and hence the final results could be dependent on the order of things being compared.

The change in the order of inputs to a Min aggregation over an array can result in different outputs based on the order of inputs, for example, for the input arrays [1,2], [3, 4], [3, Null],[3, Null]], the minimum is [1,2]. However, for the input arrays [3, Null], [3, Null], [1,2], [3, 4], the minimum is indeterminate because we cannot compare [3, Null] with [3, Null].

To solve this issue, we ended up not allowing nested nulls to ensure that the final results are deterministic since shuffles can change the order of inputs. For more information about this issue, see https://github.com/facebookincubator/velox/pull/6723.

To better solve this and still support nested nulls, I propose a new method for comparing values for the purpose of finding Min/Max that is deterministic across different input ordering.

Let's think about finding the min value as an aggregation where we consume every input in the incoming stream. Instead of stopping when a null is found, we would continue as follows:
 - The idea is that when a partial aggregation of min is indeterminate, we store one of the elements that failed the comparison in the accumulator (either one of the compared elements would work in this case) We call IndeterminiteMinSketch.
 - Then, if we compare that IndeterminiteMinSketch with any other element that is deterministically less, it would also be less than any other element that when compared with IndeterminiteMinSketch results in an indeterminate outcome. Hence, we can move on from the indeterminate state of the accumulator to a determinate one.
- If we compare that IndeterminiteMinSketch with any other element that is deterministically larger, it would also be larger than any other element that when compared with IndeterminiteMinSketch results in an indeterminate outcome. Hence, we can stay in the indeterminate state with that IndeterminiteMinSketch.

More concrete explanation with examples:
 - Basically, the accumulator would be a tuple {bool: indeterminate, X: value}
  - For example, if the input is [3, Null], [3, Null], [1,2], the accumulator would be updated as follows:
    - acc = (true, [3, Null])
    - acc = (false, [3, Null])
    - acc = (true, [1, 2])
- What happens if we see a larger value after an indeterminate? We simply ignore it. For example, if the accumulator is (false, [3, Null]) and we saw [4, Null], then since we are looking for the minimum and we failed to compare a number less than [4, Null], the result remains (false, [3, Null]).

Proof: need to add it still if people like the idea.

**Comment by laithsakka** [comment]
cc @mbasmanova  @duanmeng

**Comment by laithsakka** [comment]
This does not only solve the deterministic issue it also give more accurate  results for min/max. computation instead of stopping at first seen null with failure.
ex: I can find the min across [3, null], [3, null][1,2] why would presto fail?

**Comment by mbasmanova** [comment]
@laithsakka This is very interesting idea. I think I understand the intuition behind this approach, but I don't fully understand the details. Will chat with your over VC.

**Comment by mbasmanova** [comment]
CC: @duanmeng

**Comment by duanmeng** [comment]
@laithsakka Looks great. IMHO we could implement this by modifying the `compareNulls` logic that returns 1 or -1 according to the `CompareFlag::nullFirst` flag or something else when comparing a null with a non-null and returning `std::nullopt` when comparing a null with a null.

**Comment by duanmeng** [comment]
But for `[1,2], [2, 4], [3, null],[3, null]`, the result of min would be non-deterministic,
- the result is [1, 2] if they are divided by two streams, `([1, 2], [3, null])` and `([2, 4], [3, null])`
- the result is **non-deterministic** if they are divided by two streams, `([1, 2], [2, 4])` and `([3, null],[3, null])` that the processing of the second stream would throw null vs null exceptions.
